src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+const BASE64_KEYSTR = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
+
+function normalizeEncodingArgs(data: any, alternateCodec: any): [ string, ByteBuffer ]{
+	data = String(data);
+
+	return [ data, <ByteBuffer> {
+		toString: function (codec?: any): string {
+			return (codec || alternateCodec).decode(this);
+		},
+		toJSON: function (): string {
+			return JSON.stringify(Array.prototype.slice.call(this));
+		},
+		length: data.length
+	}];
+}
+
+function validateDecodingArgs(data: any) {
+	if (data == null) {
+		return false;
+	}
+
+	if (data.length == null) {
+		throw new TypeError('Argument must have a length property');
+	}
+
+	return true;
+}
+
+function validateUtf8EncodedCodePoint(codePoint: number): void {
+	if ((codePoint & 0xC0) !== 0x80) {
+		throw Error('Invalid continuation byte');
+	}
+}
+
+function decodeUtf8EncodedCodePoint(codePoint: number, validationRange: number[] = [0, Infinity], checkSurrogate?: boolean): string {
+	if (codePoint < validationRange[0] || codePoint > validationRange[1]) {
+		throw Error('Invalid continuation byte');
+	}
+
+	if (checkSurrogate && codePoint >= 0xD800 && codePoint <= 0xDFFF) {
+		throw Error('Surrogate is not a scalar value');
+	}
+
+	let encoded = '';
+
+	if (codePoint > 0xFFFF) {
+		codePoint = codePoint - 0x010000;
+		encoded += String.fromCharCode(codePoint >>> 0x10 & 0x03FF | 0xD800);
+		codePoint = 0xDC00 | codePoint & 0x03FF;
+	}
+
+	encoded += String.fromCharCode(codePoint);
+
+	return encoded;
+}
+
+export interface ByteBuffer {
```

Comment:
@jason0x43 may have more detailed thoughts on this, but it seems like it'd be far more useful (though probably more difficult) to be able to return something that is useful in the respective platform this is used.

At the *very* least, it seems like it'd be more useful for this to actually be an array, not just an array-like.